aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+//! Key Transport: KEMs, including support for Kyber.
+//!
+//! # Example
+//!
+//! Note that this example uses Kyber-512, but other algorithms can be used
+//! exactly the same way, just substituting
+//! `KemAlgorithm::<desired_algorithm_here>` for `KemAlgorithm::KYBER512_R3`.
+//!
+//! ```
+//! use aws_lc_rs::key_transport::*;
+//!
+//! let priv_key = KemPrivateKey::generate(KemAlgorithm::KYBER512_R3)?;
+//!
+//! // Generate private key bytes to possibly save for later decapsulation
+//! let privkey_raw_bytes = priv_key.get_raw_bytes();
+//!
+//! let pub_key = priv_key.compute_public_key()?;
+//!
+//! // Get the public key bytes to send to bob through some encoded
+//! // protocol message.
+//! let pub_key_bytes = pub_key.get_raw_bytes();
+//!
+//! let mut ciphertext: Vec<u8> = vec![];
+//!
+//! let retrieved_pub_key = KemPublicKey::from_raw_bytes(KemAlgorithm::KYBER512_R3, pub_key_bytes)?;
+//! let bob_result = retrieved_pub_key.encapsulate(|ct, ss| {
+//!     ciphertext.extend_from_slice(ct);
+//!     // In a real application, we'd apply a KDF to the shared secret and the
+//!     // public keys (as recommended in RFC 7748) and then derive session
+//!     // keys from the result. We omit all that here.
+//!     Ok(())
+//! });
+//! 
+//! // Retrieve private key from stored raw bytes
+//! let retrieved_priv_key = KemPrivateKey::from_raw_bytes(KemAlgorithm::KYBER512_R3, privkey_raw_bytes)?;
+//!
+//! let alice_result = retrieved_priv_key.decapsulate(&mut ciphertext, |ss| {
+//!     // In a real application, we'd apply a KDF to the shared secret and the
+//!     // public keys (as recommended in RFC 7748) and then derive session
+//!     // keys from the result. We omit all that here.
+//!     Ok(())
+//! });
+//! 
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+use crate::{error::{Unspecified, KeyRejected}, ptr::LcPtr, ptr::{DetachableLcPtr, Pointer}};
+use std::{os::raw::c_int};
+use std::ptr::null_mut;
+use aws_lc::{
+    EVP_PKEY, NID_KYBER512_R3, EVP_PKEY_keygen, EVP_PKEY_CTX_new_id, EVP_PKEY_keygen_init,
+    EVP_PKEY_KEM, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new,
+    EVP_PKEY_encapsulate, EVP_PKEY_decapsulate, EVP_PKEY_get_raw_public_key, EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_get_raw_private_key
+};
+use zeroize::Zeroize;
+
+
+const KYBER512_SECRETKEYBYTES: usize = 1632;
+const KYBER512_PUBLICKEYBYTES: usize = 800;
+const KYBER512_CIPHERTEXTBYTES: usize = 768;
+const KYBER512_BYTES: usize = 32;
```

Comment:
I stole these from the aws-lc source code for Kyber. I agree with the name changes. I'll double check the numbers

Reply:
Also you don't need to duplicate these, if you call keygen/encapsulate/decapsulate with null buffers we set the expected size for you. This could make it easier to add more KEMs in the future. https://github.com/aws/aws-lc/blob/main/include/openssl/evp.h#L723-L726

Reply:
Yeah, I thought maybe it would be better for performance if we're making less C FFI calls. Probably negligible though and worth getting rid of the constants.

Reply:
We use constants for fixed key size everywhere else in the package. I think they're fine as long as the logic encapsulate/decapsulate/etc is obtaining these sizes via a call to the `KemAlgorithm` and not directly.

Reply:
Lengths are as specified on the crystals website: [https://pq-crystals.org/kyber/](https://pq-crystals.org/kyber/)